File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {

        //using minHeap PriorityQueue from java 
        PriorityQueue<ListNode> pointers = new PriorityQueue<>((a, b) -> a.val - b.val);
        ListNode res = new ListNode(0);
        ListNode cur = res;
        //keep a tab on each node, store and remove from priority queue as needed

        for(int i = 0; i<lists.length; i++){
            if(lists[i] != null){
            pointers.add(lists[i]);
            }
        }

        //there exists some pointer that has not reached null
        while(!pointers.isEmpty()){
            ListNode polled = pointers.poll();
            cur.next = polled;
            cur = cur.next;
            if(polled.next != null){
                pointers.add(polled.next);
            }
        }
        return res.next;
    }
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:

        if not head:
            return None
        
        node = head
        length = 0

        while node:
            length += 1
            node = node.next

        headTemp = head
        cycles = length // k
        if cycles == 0:                 
            return head
        last = None
        for i in range(cycles):
            node = headTemp
            dummy = None                
            size = k

            while size:
                nextNode = node.next
                node.next = dummy
                dummy = node
                node = nextNode
                size -= 1

            if i == 0:
                res = dummy

            if last:                   
                last.next = dummy
            last = headTemp            
            headTemp = node             

        last.next = headTemp           
        return res


            

        
            

        